File: backend/app/services/voice.py
import logging

import httpx

from app.config import settings

log = logging.getLogger(__name__)

RETELL_URL = "https://api.retellai.com/v2/create-phone-call"


async def make_call(
    retell_agent_id: str,
    phone_number: str,
    metadata: dict | None = None,
    from_number: str | None = None,
    conversation_flow_id: str | None = None,
    retell_llm_dynamic_variables: dict | None = None,
) -> dict:
    if not phone_number.startswith("+"):
        raise ValueError(f"phone_number must be E.164 format (e.g. +15551234567), got: {phone_number}")

    body: dict = {
        "agent_id": retell_agent_id,
        "to_number": phone_number,
        "metadata": metadata or {},
    }
    if conversation_flow_id:
        body["conversation_flow_id"] = conversation_flow_id
    if retell_llm_dynamic_variables:
        body["retell_llm_dynamic_variables"] = retell_llm_dynamic_variables

    _from = from_number or settings.retell_from_number
    if _from:
        body["from_number"] = _from

    async with httpx.AsyncClient(timeout=30) as client:
        resp = await client.post(
            RETELL_URL,
            headers={
                "Authorization": f"Bearer {settings.retell_api_key}",
                "Content-Type": "application/json",
            },
            json=body,
        )

    if not resp.is_success:
        raise RuntimeError(f"[retell] status={resp.status_code} body={resp.text}")

    log.info("[voice] call initiated to=%s agent=%s", phone_number, retell_agent_id)
    return resp.json()

File: backend-rust/src/handlers/voice.rs
//! Retell AI voice call handler. Payload carries the Retell agent_id,
//! optional conversation_flow_id, and dynamic_variables. The Retell API key
//! is redeemed via credential ref.

use crate::credentials;
use crate::handlers::common;
use crate::models::{ActionCommand, ExecutionResult};
use serde_json::{json, Value};

const RETELL_URL: &str = "https://api.retellai.com/v2/create-phone-call";

pub async fn handle_voice(command: &ActionCommand) -> ExecutionResult {
    let retell_agent_id = common::s(command, "retell_agent_id");
    let phone = command.lead.phone.clone().unwrap_or_default();
    if retell_agent_id.is_empty() || phone.is_empty() {
        return common::fail(command, "retell_agent_id or lead.phone missing", false);
    }
    if !phone.starts_with('+') {
        return common::fail(command, format!("phone must be E.164 (got {phone})"), false);
    }

    let cred_ref = match command.credential_ref.as_ref() {
        Some(r) => r.clone(),
        None => return common::fail(command, "voice command missing credential_ref", false),
    };
    let api_key = match credentials::redeem_field(&cred_ref, "api_key").await {
        Ok(Some(k)) => k,
        Ok(None) => return common::fail(command, "credential bundle missing api_key", false),
        Err(e) => return common::fail(command, e, true),
    };

    let mut body = json!({
        "agent_id": retell_agent_id,
        "to_number": phone,
        "metadata": {
            "lead_id": command.lead.id.to_string(),
            "campaign_id": command.lead.campaign_id.to_string(),
        }
    });
    if let Some(flow) = common::opt_s(command, "conversation_flow_id") {
        body["conversation_flow_id"] = Value::String(flow);
    }
    if let Some(from_number) = common::opt_s(command, "from_number") {
        body["from_number"] = Value::String(from_number);
    }
    if let Some(vars) = command.payload.get("retell_llm_dynamic_variables") {
        body["retell_llm_dynamic_variables"] = vars.clone();
    }

    let client = reqwest::Client::new();
    let resp = client
        .post(RETELL_URL)
        .bearer_auth(&api_key)
        .json(&body)
        .send()
        .await;
    credentials::release(&cred_ref).await;

    match resp {
        Ok(r) if r.status().is_success() => {
            let j: Value = r.json().await.unwrap_or(json!({}));
            let call_id = j.get("call_id").and_then(|v| v.as_str()).unwrap_or("").to_string();
            common::ok(
                command,
                json!({"provider": "retell", "call_id": call_id}),
                Some("call_made"),
                json!({}),
            )
        }
        Ok(r) => {
            let s = r.status();
            let t = r.text().await.unwrap_or_default();
            common::fail(command, format!("retell HTTP {s}: {}", t.chars().take(200).collect::<String>()), s.is_server_error() || s.as_u16() == 429)
        }
        Err(e) => common::fail(command, format!("retell network: {e}"), true),
    }
}
